go/enclave/gas/oracle.go

```diff
@@ -44,45 +67,92 @@ func (o *oracle) EstimateL1StorageGasCost(tx *types.Transaction, block *types.He
 	}
 
-	return o.calculateL1Cost(block, header, encoded)
+	return o.calculateL1Cost(nil, header, encoded)
+}
+
+func (o *oracle) SubmitL1Block(ctx context.Context, headBlock *types.Header) error {
+	o.headMutex.Lock()
+	defer o.headMutex.Unlock()
+	o.headBlock = headBlock
+	calculateBlobs := headBlock.ExcessBlobGas == nil
+
+	baseFeeSum := big.NewInt(0)
+	blobFeeSum := big.NewInt(0)
+	count := 0
+
+	b := headBlock
+	var err error
+	for ; count < MovingAverageWindow; count++ {
+		baseFeeSum = baseFeeSum.Add(baseFeeSum, b.BaseFee)
+		if calculateBlobs {
+			blobFeeSum = blobFeeSum.Add(blobFeeSum, eip4844.CalcBlobFee(o.l1ChainCfg, b))
+		}
+		b, err = o.storage.FetchBlock(ctx, b.ParentHash)
+		if err != nil {
+			break
+		}
+	}
+
+	baseFeeMA, blobFeeMA := big.NewInt(0), big.NewInt(0)
+	if count > 0 {
+		baseFeeMA = new(big.Int).Div(baseFeeSum, big.NewInt(int64(count)))
+		if calculateBlobs {
+			blobFeeMA = new(big.Int).Div(blobFeeSum, big.NewInt(int64(count)))
+		}
+	}
+
+	blockNum := headBlock.Number.Uint64()
+	o.baseFeeMA[blockNum] = baseFeeMA
+	if blockNum > MaxHistoricMA {
+		// cleanup entries older than MaxHistoricMA
+		delete(o.baseFeeMA, blockNum-MaxHistoricMA)
+		delete(o.blobFeeMA, blockNum-MaxHistoricMA)
+	}
+	o.blobFeeMA[headBlock.Number.Uint64()] = blobFeeMA
+
+	return nil
 }
 
 // calculateL1Cost - Calculates the L1 cost as a multiple of the L2 base fee.
 // it takes into account the share of the blob cost and the share of the L1 TX cost - which submits and stores the rollup header.
-func (o *oracle) calculateL1Cost(l1Block *types.Header, l2Batch *common.BatchHeader, encodedTx []byte) (*big.Int, error) {
-	totalCost := big.NewInt(0)
+func (o *oracle) calculateL1Cost(block *types.Header, l2Batch *common.BatchHeader, encodedTx []byte) (*big.Int, error) {
+	o.headMutex.RLock()
+	defer o.headMutex.RUnlock()
 
-	// If the l1Block does not have excess blob gas, we can't estimate the cost
-	if l1Block.ExcessBlobGas == nil {
-		return totalCost, nil
+	if block == nil {
+		block = o.headBlock
 	}
 
 	// 1. Calculate the cost of including the tx in a blob
 	// price in Wei for a single unit of blob
-	// todo - use a moving average for the L1 blob fee
-	// todo pass in the L1 chain config
-	blobFeePerByte := eip4844.CalcBlobFee(o.l1ChainCfg, l1Block)
+	shareOfBlobCost := big.NewInt(0)
 	txL1Size := CalculateL1Size(encodedTx)
-	shareOfBlobCost := big.NewInt(0).Mul(txL1Size, blobFeePerByte)
+	bl := o.blobFeeMA[block.Number.Uint64()]
+	if bl != nil && bl.Sign() > 0 {
+		shareOfBlobCost = big.NewInt(0).Mul(txL1Size, o.blobFeeMA[block.Number.Uint64()])
+	}
 
 	// 2. Estimate how much this tx should absorb from the L1 tx cost that submits the rollup
 	shareOfL1TxGas := big.NewInt(L1TxGas / TxsPerRollup)
-	// todo - use a moving average for the L1 base fee
-	shareOfL1TxCost := big.NewInt(0).Mul(shareOfL1TxGas, l1Block.BaseFee)
+	shareOfL1TxCost := big.NewInt(0)
+	bf := o.baseFeeMA[block.Number.Uint64()]
+	if bf != nil && bf.Sign() > 0 {
```

Comment:
what's the sign() > 0

Reply:
should probably be encapsulated as isPositive := bf.Sign() > 0 if that's the intention

Reply:
it's checking that it's not `0`